clients/vscode-hlasmplugin/src/hlasmExternalFilesFtp.ts

```diff
@@ -0,0 +1,234 @@
+                await this.mutex.locked(async () => {
+                    const connection = await this.getConnInfo();
+
+                    await client.access(translateConnectionInfo(connection));
+
+                    client.parseList = (rawList: string): ftp.FileInfo[] => {
+                        return rawList.split(/\r?\n/).slice(1).filter(x => !/^\s*$/.test(x)).map(value => new ftp.FileInfo(value.split(/\s/, 1)[0]));
+                    };
+
+                    this.activeConnectionInfo = connection;
+                });
+
+                return client;
+            }
+            catch (e) {
+                this.activeConnectionInfo = null;
+
+                if (e instanceof ftp.FTPError) {
+                    vscode.window.showErrorMessage(e.message);
+                    continue;
+                }
+
+                client.close();
+
+                throw e;
+            }
+        }
+    }
+
+    async listMembersImpl(client: ftp.Client, dataset: string): Promise<string[] | null> {
+        try {
+            await checkedCommand(client, 'TYPE A');
+            checkResponse(await client.cd(`'${dataset}'`));
+            const list = await client.list();
+            return list.map(x => x.name);
+        }
+        catch (e) {
+            if (e instanceof ftp.FTPError && e.code == 550)
+                return null;
+            throw e;
+        }
+    }
+
+    async listMembers(args: DatasetUriDetails): Promise<string[] | null> {
+        const client = await this.getConnectedClient();
+        return this.listMembersImpl(client, args.dataset).finally(() => { this.requestEndHandle(client) });
+    }
+
+    async readMemberImpl(client: ftp.Client, dataset: string, member: string): Promise<string | null> {
+        try {
+            const buffer = new FBWritable();
+            buffer.on('error', err => { throw err });
+
+            await checkedCommand(client, 'TYPE I');
+            checkResponse(await client.downloadTo(buffer, `'${dataset}(${member})'`));
+
+            return buffer.getResult();
+        }
+        catch (e) {
+            if (e instanceof ftp.FTPError && e.code == 550)
+                return null;
+            throw e;
+        }
+    }
+    async readMember(args: DatasetUriDetails): Promise<string | null> {
+        const client = await this.getConnectedClient();
+        return this.readMemberImpl(client, args.dataset, args.member).finally(() => { this.requestEndHandle(client); });
+    }
+
+    parseArgs(path: string, purpose: ExternalRequestType): ClientUriDetails {
+        const [dataset, member] = path.split('/').slice(1).map(x => x.toUpperCase());
+
+        if (!dataset || !/^(?:[A-Z$#@][A-Z$#@0-9]{1,7})(?:\.(?:[A-Z$#@][A-Z$#@0-9]{1,7}))*$/.test(dataset) || dataset.length > 44) return null;
+        switch (purpose) {
+            case ExternalRequestType.read_file:
+                if (!member || !/^[A-Z$#@][A-Z$#@0-9]{1,7}(?:\..*)?$/.test(member)) return null; // ignore extension
```

Comment:
The member name length is checked by the regex, the extra check on the data set name is needed because the regex only validates that the data set name consists of the correct/valid qualifiers, but then the global length property needs to be validated as well.